olive/passes/onnx/distribute_experts.py

```diff
@@ -0,0 +1,226 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+
+import argparse
+import json
+import numpy
+import onnx
+import os
+import pprint
+import sys
+
+from copy import deepcopy
+from google.protobuf.json_format import MessageToDict
+from google.protobuf.message import Message
+from olive.model import ONNXModel
+from olive.passes import Pass
+from olive.passes.pass_config import PassConfigParam
+from onnxruntime.transformers.onnx_model import OnnxModel
+from pathlib import Path
+from typing import Any, Dict
+
+
+_domain = 'com.microsoft'
+_json_separators = (',', ': ')
+_expert_pattern = [
+  'Slice', 'Concat', 'MatMul', 'Cast', 'Relu', 'MatMul', 'Cast',
+  'Slice', 'Mul', 'Div', 'Add', 'Gather', 'Shape', 'Reshape', 'Reshape'
+]
+
+
+def _dump_graph(node: Message, filepath: str):
+    with open(filepath, 'wt') as strm:
+        json.dump(MessageToDict(node), fp=strm, indent=2, sort_keys=True, separators=_json_separators)
+        strm.flush()
+
+
+def _iterate_experts(model: OnnxModel, producers: dict):
+  for node in model.get_nodes_by_op_type(_expert_pattern[0]):
+      path = model.match_parent_path(node, _expert_pattern[1:], output_name_to_node=producers)
+      if path:
+          path.reverse()
+          path.append(node)
+          yield path
+
+
+def _create_node_name(model: OnnxModel, op_type: str, prefix_a: str, prefix_b: str):
+    prefix = ''
+    last_slash = -1
+    for i in range(0, min(len(prefix_a), len(prefix_b))):
+        if prefix_a[i] == prefix_b[i]:
+            prefix += prefix_a[i]
+            if prefix_a[i] == '/':
+                last_slash = i
+        else:
+            break
+
+    if last_slash > 0:
+        prefix = prefix[:last_slash+1]
+
+    if not prefix.endswith('/'):
+        prefix += '/'
+    prefix += op_type
+
+    return model.create_node_name(op_type, prefix)
+
+
+def _create_ranked_model(model: OnnxModel, nodes: dict, producers: dict, consumers: dict, experts: list, world_size: int):
+    for expert in experts:
+        div_node = nodes[expert[5]]
+        div_node_oname = div_node.output[0]
+        mul_nodes = [n for n in consumers[div_node_oname] if n.op_type == 'Mul']
+
+        reshape_node = nodes[expert[1]]
+        reshape_node_oname = reshape_node.output[0]
+        slice_nodes = [n for n in consumers[reshape_node_oname] if n.op_type == 'Slice']
+
+        concat_node = nodes[expert[13]]
+        matmul_nodes = [producers[iname] for iname in concat_node.input]
+
+        num_experts = len(consumers[div_node_oname])
+        experts_per_rank = int(num_experts / world_size)
+
+        for i in range(experts_per_rank, num_experts):
+            mul_nodes[i].input.remove(div_node_oname)
```

Comment:
If there are 16 experts and world size is 4 then does this mean all ranks will get first 4 experts?

Reply:
Each rank should get different set of 4 experts.